File: backend/app/schemas/market.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class EmotionSnapshot(BaseModel):
    """市场情绪快照（market.load_emotion 产出的固定结构）。"""

    trade_date: str
    max_limit_times: int
    max_board_vt_symbol: str
    linked_board_count: int
    linked_board_vt_symbols: list[str] = Field(default_factory=list)
    updated_at: str


class MarketOverview(BaseModel):
    redis_available: bool
    quote_count: int
    updated_at: str | None = None
    is_trading: bool = False
    emotion: EmotionSnapshot | None = None
    emotion_cycle: EmotionCycleOut | None = None
    ranks_available: list[str] = Field(default_factory=list)


class EmotionThresholdsOut(BaseModel):
    recession_limit_down: int
    ice_max_boards: int
    ice_limit_down: int
    ice_up_ratio_max: float
    climax_ladder_depth: int
    climax_limit_up: int
    divergence_limit_up_min: int
    divergence_limit_spread: int
    startup_max_boards: int
    startup_limit_up: int
    amount_floor_yuan: float
    recession_break_rate: float
    fear_greed_overheat: float
    hysteresis_enabled: bool
    is_default: bool = True


class EmotionThresholdsPut(BaseModel):
    recession_limit_down: int | None = None
    ice_max_boards: int | None = None
    ice_limit_down: int | None = None
    ice_up_ratio_max: float | None = None
    climax_ladder_depth: int | None = None
    climax_limit_up: int | None = None
    divergence_limit_up_min: int | None = None
    divergence_limit_spread: int | None = None
    startup_max_boards: int | None = None
    startup_limit_up: int | None = None
    amount_floor_yuan: float | None = None
    recession_break_rate: float | None = None
    fear_greed_overheat: float | None = None
    hysteresis_enabled: bool | None = None


class EmotionCycleInputs(BaseModel):
    """情绪周期判定输入（build_emotion_cycle 的 inputs 子结构）。"""

    limit_up_count: int = 0
    limit_down_count: int = 0
    up_ratio: float = 0.0
    total_amount: float = 0.0
    max_limit_times: int = 0
    limit_ladder_depth: int = 0
    prev_leader_limit_down: bool = False
    limit_break_rate: float | None = None
    index_above_ma5: bool | None = None
    fear_greed_index: float = 0.0
    fear_greed_source: str = ""
    sample_size: int = 0


class EmotionCycleOut(BaseModel):
    stage: str
    stage_label: str
    position_factor: float
    position_pct_min: float
    position_pct_max: float
    allow_new_positions: bool
    allowed_modes: list[str] = Field(default_factory=list)
    allowed_mode_labels: list[str] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    source: str = ""
    trade_date: str | None = None
    raw_stage: str | None = None
    inputs: EmotionCycleInputs = Field(default_factory=EmotionCycleInputs)


class RankRow(BaseModel):
    rank: int
    symbol: str
    exchange: str
    vt_symbol: str
    tf_symbol: str
    name: str = ""
    score: float
    last_price: float | None = None
    change_pct: float | None = None
    turnover_rate: float | None = None
    amount: float | None = None
    volume_ratio: float | None = None
    limit_times: float | None = None


class SectorFlowRow(BaseModel):
    trade_date: str
    sector_kind: str
    sector_id: str
    name: str
    change_pct: float
    net_flow_yi: float
    flow_source: str = ""


class SectorIntradayPoint(BaseModel):
    bucket_time: str
    clock_minutes: int
    net_flow_yi: float
    change_pct: float


class RadarCardOut(BaseModel):
    card_id: str
    title: str
    subtitle: str = ""
    source: str  # cache | synthesized
    computed_at: str = ""
    empty_message: str = ""
    # 异构卡片行：不同 card_id 行结构不同（板块/连板/涨幅榜/放量/龙头…），前端按 card_id 渲染，故保留开放结构。
    rows: list[dict[str, Any]] = Field(default_factory=list)


class RadarResonanceEntry(BaseModel):
    vt_symbol: str
    name: str = ""
    card_count: int
    card_titles: list[str] = Field(default_factory=list)
    resonance_score: float
    change_pct: float | None = None
    last_price: float | None = None
    seal_time_label: str = ""


class RadarResonanceOut(BaseModel):
    min_cards: int
    top_n: int
    total: int
    entries: list[RadarResonanceEntry] = Field(default_factory=list)


class RadarResonanceWeightItem(BaseModel):
    card_id: str
    title: str
    weight: float
    default_weight: float


class RadarResonanceWeightsOut(BaseModel):
    items: list[RadarResonanceWeightItem] = Field(default_factory=list)
    weights: dict[str, float] = Field(default_factory=dict)


class RadarResonanceWeightsPut(BaseModel):
    weights: dict[str, float] = Field(default_factory=dict)


class RadarHorizonRow(BaseModel):
    vt_symbol: str
    name: str = ""
    resonance_score: float = 0
    card_count: int = 0
    card_titles: list[str] = Field(default_factory=list)
    change_pct: float | None = None
    last_price: float | None = None
    seal_time_label: str = ""


class RadarHorizonOut(BaseModel):
    variant: str = "default"
    strategy_key: str = ""
    computed_at: str | None = None
    scanned_total: int = 0
    refined_total: int = 0
    rows: list[RadarHorizonRow] = Field(default_factory=list)
    empty: bool = True
    label: str = "启发式展望（基于共振）"


class RadarPredictRow(BaseModel):
    vt_symbol: str
    name: str = ""
    predict_score: float = 0
    resonance_score: float = 0
    card_count: int = 0
    card_titles: list[str] = Field(default_factory=list)
    change_pct: float | None = None
    last_price: float | None = None
    seal_time_label: str = ""
    reasons: list[str] = Field(default_factory=list)


class RadarPredictOut(BaseModel):
    variant: str = "default"
    model_label: str = ""
    computed_at: str | None = None
    scanned_total: int = 0
    refined_total: int = 0
    kline_missing: int = 0
    rows: list[RadarPredictRow] = Field(default_factory=list)
    empty: bool = True
    label: str = "规则预测（共振+可解释加分）"


class LimitListRow(BaseModel):
    trade_date: str = ""
    vt_symbol: str = ""
    ts_code: str = ""
    name: str = ""
    limit_times: float = 0
    first_time: str = ""
    last_time: str = ""
    fd_amount: float = 0
    open_times: float = 0
    strth: float = 0
    updated_at: str = ""
    seal_time_score: float = 0
    seal_time_label: str = ""


class LimitListOut(BaseModel):
    trade_date: str = ""
    total: int = 0
    rows: list[LimitListRow] = Field(default_factory=list)


class PlanDraftRequest(BaseModel):
    top_n: int | None = None
    trade_date: str | None = None


class PlanDraftSymbol(BaseModel):
    vt_symbol: str
    name: str = ""


class PlanDraftOut(BaseModel):
    plan_id: str
    trade_date: str
    status: str
    emotion_expected: str
    symbol_count: int
    symbols: list[PlanDraftSymbol] = Field(default_factory=list)
    replaced: bool

File: frontend/src/api/market.ts
import { api } from './client'

export type MarketOverview = {
  redis_available: boolean
  quote_count: number
  updated_at: string | null
  is_trading: boolean
  emotion: {
    trade_date: string
    max_limit_times: number
    max_board_vt_symbol: string
    linked_board_count: number
    linked_board_vt_symbols: string[]
    updated_at: string
  } | null
  emotion_cycle: EmotionCycle | null
  ranks_available: string[]
}

export type EmotionCycle = {
  stage: string
  stage_label: string
  position_factor: number
  position_pct_min: number
  position_pct_max: number
  allow_new_positions: boolean
  allowed_modes: string[]
  allowed_mode_labels: string[]
  warnings: string[]
  source: string
  trade_date: string | null
  inputs: {
    limit_up_count?: number
    limit_down_count?: number
    up_ratio?: number
    total_amount?: number
    max_limit_times?: number
    limit_ladder_depth?: number
    prev_leader_limit_down?: boolean
    sample_size?: number
    fear_greed_index?: number
    index_above_ma5?: boolean | null
    limit_break_rate?: number | null
  }
}

export type RankRow = {
  rank: number
  symbol: string
  exchange: string
  vt_symbol: string
  tf_symbol: string
  name: string
  score: number
  last_price: number | null
  change_pct: number | null
  turnover_rate: number | null
  amount: number | null
  volume_ratio: number | null
  limit_times: number | null
}

export type SectorFlowRow = {
  trade_date: string
  sector_kind: string
  sector_id: string
  name: string
  change_pct: number
  net_flow_yi: number
  flow_source: string
}

export type RadarCard = {
  card_id: string
  title: string
  subtitle: string
  source: string
  computed_at: string
  empty_message: string
  rows: Record<string, unknown>[]
}

export type RadarResonanceEntry = {
  vt_symbol: string
  name: string
  card_count: number
  card_titles: string[]
  resonance_score: number
  change_pct: number | null
  last_price: number | null
  seal_time_label?: string
}

export type LimitListRow = {
  trade_date: string
  vt_symbol: string
  ts_code: string
  name: string
  limit_times: number
  first_time: string
  last_time: string
  fd_amount: number
  open_times: number
  strth: number
  updated_at: string
  seal_time_score: number
  seal_time_label: string
}

export type LimitListOut = {
  trade_date: string
  total: number
  rows: LimitListRow[]
}

export type RadarResonance = {
  min_cards: number
  top_n: number
  total: number
  entries: RadarResonanceEntry[]
}

export type RadarHorizon = {
  variant: string
  strategy_key: string
  computed_at: string | null
  scanned_total: number
  refined_total: number
  rows: Array<{
    vt_symbol: string
    name: string
    resonance_score: number
    card_count: number
    card_titles: string[]
    change_pct: number | null
    last_price: number | null
    seal_time_label?: string
  }>
  empty: boolean
  label: string
}

export type RadarPredict = {
  variant: string
  model_label: string
  computed_at: string | null
  scanned_total: number
  refined_total: number
  kline_missing: number
  rows: Array<{
    vt_symbol: string
    name: string
    predict_score: number
    resonance_score: number
    card_count: number
    card_titles: string[]
    change_pct: number | null
    last_price: number | null
    seal_time_label?: string
    reasons: string[]
  }>
  empty: boolean
  label: string
}

export type ResonanceWeightItem = {
  card_id: string
  title: string
  weight: number
  default_weight: number
}

export type ResonanceWeights = {
  items: ResonanceWeightItem[]
  weights: Record<string, number>
}

export type EmotionThresholds = {
  recession_limit_down: number
  ice_max_boards: number
  ice_limit_down: number
  ice_up_ratio_max: number
  climax_ladder_depth: number
  climax_limit_up: number
  divergence_limit_up_min: number
  divergence_limit_spread: number
  startup_max_boards: number
  startup_limit_up: number
  amount_floor_yuan: number
  recession_break_rate: number
  fear_greed_overheat: number
  hysteresis_enabled: boolean
  is_default: boolean
}

export type EmotionThresholdsPatch = Partial<Omit<EmotionThresholds, 'is_default'>>

export type PlanDraftOut = {
  plan_id: string
  trade_date: string
  status: string
  emotion_expected: string
  symbol_count: number
  symbols: { vt_symbol: string; name?: string }[]
  replaced: boolean
}

export const marketApi = {
  overview: () => api<MarketOverview>('/api/v1/market/overview'),
  emotionCycle: () => api<EmotionCycle>('/api/v1/market/emotion-cycle'),
  ranks: (field = 'change_pct', topN = 50) =>
    api<RankRow[]>(`/api/v1/market/ranks?field=${encodeURIComponent(field)}&top_n=${topN}`),
  sectorDates: () => api<string[]>('/api/v1/sectors/dates'),
  sectorFlow: (
    opts: { kind?: string; trade_date?: string; sort?: string; limit?: number } = {},
  ) => {
    const q = new URLSearchParams()
    q.set('kind', opts.kind || 'industry')
    q.set('sort', opts.sort || 'net_flow_yi')
    q.set('limit', String(opts.limit || 50))
    if (opts.trade_date) q.set('trade_date', opts.trade_date)
    return api<SectorFlowRow[]>(`/api/v1/sectors/flow?${q}`)
  },
  radarCards: () => api<RadarCard[]>('/api/v1/radar/cards'),
  radarHorizon: () => api<RadarHorizon>('/api/v1/radar/horizon'),
  radarPredict: () => api<RadarPredict>('/api/v1/radar/predict'),
  radarResonance: (opts: { top_n?: number; min_cards?: number } = {}) => {
    const q = new URLSearchParams()
    q.set('top_n', String(opts.top_n ?? 20))
    q.set('min_cards', String(opts.min_cards ?? 2))
    return api<RadarResonance>(`/api/v1/radar/resonance?${q}`)
  },
  resonanceWeights: () => api<ResonanceWeights>('/api/v1/radar/resonance/weights'),
  putResonanceWeights: (weights: Record<string, number>) =>
    api<ResonanceWeights>('/api/v1/radar/resonance/weights', {
      method: 'PUT',
      body: JSON.stringify({ weights }),
    }),
  limitList: (tradeDate?: string) => {
    const q = tradeDate ? `?trade_date=${encodeURIComponent(tradeDate)}` : ''
    return api<LimitListOut>(`/api/v1/market/limit-list${q}`)
  },
  emotionThresholds: () => api<EmotionThresholds>('/api/v1/market/emotion-cycle/thresholds'),
  putEmotionThresholds: (body: EmotionThresholdsPatch) =>
    api<EmotionThresholds>('/api/v1/market/emotion-cycle/thresholds', {
      method: 'PUT',
      body: JSON.stringify(body),
    }),
  resetEmotionThresholds: () =>
    api<EmotionThresholds>('/api/v1/market/emotion-cycle/thresholds/reset', {
      method: 'POST',
    }),
  createPlanDraft: (body: { top_n?: number; trade_date?: string } = {}) =>
    api<PlanDraftOut>('/api/v1/radar/plan-draft', {
      method: 'POST',
      body: JSON.stringify(body),
    }),
}
